subgraphs/rocket-pool/src/entities/token.ts

```diff
@@ -0,0 +1,121 @@
+import { Address, BigDecimal, BigInt, log } from "@graphprotocol/graph-ts";
+import { Token } from "../../generated/schema";
+import { ERC20 } from "../../generated/rocketVault/ERC20";
+import { RETH } from "../../generated/rocketVault/RETH";
+import { getUsdPricePerToken } from "../prices";
+import { bigIntToBigDecimal } from "../utils/numbers";
+import {
+  BIGDECIMAL_ZERO,
+  ETH_ADDRESS,
+  ETH_NAME,
+  ETH_SYMBOL,
+  RETH_ADDRESS,
+  ONE_ETH_IN_WEI,
+  BIGDECIMAL_ONE,
+} from "../utils/constants";
+
+export function getOrCreateToken(
+  tokenAddress: Address,
+  blockNumber: BigInt
+): Token {
+  const tokenId = tokenAddress.toHexString();
+  let token = Token.load(tokenId);
+
+  if (!token) {
+    token = new Token(tokenId);
+
+    if (tokenAddress == Address.fromString(ETH_ADDRESS)) {
+      token.name = ETH_NAME;
+      token.symbol = ETH_SYMBOL;
+      token.decimals = 18;
+    } else {
+      token.name = fetchTokenName(tokenAddress);
+      token.symbol = fetchTokenSymbol(tokenAddress);
+      token.decimals = fetchTokenDecimals(tokenAddress) as i32;
+    }
+  }
+
+  // Optional lastPriceUSD and lastPriceBlockNumber, but used in financialMetrics
+  let rethpriceUSD: BigDecimal | null = null;
+  if (tokenAddress == Address.fromString(RETH_ADDRESS)) {
+    rethpriceUSD = getRETHpriceUSD(
+      Address.fromString(RETH_ADDRESS),
+      blockNumber
+    );
+    log.warning("[getOrCreateToken]rethPriceUSD call result: {} ", [
+      rethpriceUSD.toString(),
+    ]);
+  }
+  let price = getUsdPricePerToken(tokenAddress);
```

Comment:
Since you have available the block number here, you can check the last time the price was updated for this token so you don't fetch it multiple times on the same event or block.

Reply:
adding new return earlier makes this call necessary for non-RETH tokens.

Reply:
Have just modified this code to return early if same blockNumber. Smart workaround! Thx